distributed/shuffle/_core.py

```diff
@@ -288,7 +289,8 @@ def add_partition(
         self.raise_if_closed()
         if self.transferred:
             raise RuntimeError(f"Cannot add more partitions to {self}")
-        shards = self._shard_partition(data, partition_id)
+        with context_meter.meter("shuffle-shard-partition", func=thread_time):
```

Comment:
This cleverly excludes GIL contention. Can't pull the same trick for async instructions though.